Develop ugwp improvements

mdtoyNOAA opened the pull request:
# PR Checklist

- [x] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR. Please consult the ufs-weather-model [wiki](https://github.com/ufs-community/ufs-weather-model/wiki/Making-code-changes-in-the-UFS-weather-model-and-its-subcomponents) if you are unsure how to do this.

- [x] This PR has been tested using a branch which is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR

- [x] An Issue describing the work contained in this PR has been created either in the subcomponent(s) or in the ufs-weather-model. The Issue should be created in the repository that is most relevant to the changes in contained in the PR. The Issue and the dependent sub-component PR
are specified below.

- [x] Results for one or more of the regression tests change and the reasons for the changes are understood and explained below.


## Description

This pull request has three main components:
1)  To fix some issues with the orographic drag schemes of the "_drag_suite_" CCPP module which have caused model crashes to occur in GFSv17 p8 testing.  These issues are described in [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20).
2)  To add optional diagnostic outputs for each of the components of the drag suite for tuning purposes.  This is also described in [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20).  The optional diagnostic outputs can be selected at run time by the already-existing _ldiag_ugwp_ namelist flag.
3)  As part of the 1st component above, at the suggestion of Philip Pegion, we have added the capability to update the u and v winds in each model column based on the tendencies calculated by the mesoscale gravity wave drag and blocking schemes, and using these for the calculation of the tendencies by the small-scale gravity wave drag and turbulent orographic form drag schemes.  This sequential updating of the wind profile has been shown to improve the numerical stability.  The option to update the winds in this manner is controlled by a new logic flag _ugwp_seq_update_, which for now, is not a namelist option, but is hardwired by default to ._false_. in _GFS_typedefs.F90_.  The user can change this to ._true_. at compile time.

Also, fixed bugs in the rap, hrrr and rrfs test descriptions (changed instances of NEW_DIAGTABLE to DIAG_TABLE) and deleted extraneous line in tests/parm/rap.nml.IN (i.e., ldiag_ugwp=.false.).

Also, updated calculation of diagnostic output variables "_u-gwd_ave_" and "_v-gwd_ave_" to include surface drag stress due to small-scale GWD and Turbulent Orographic Form Drag.

Is a change of answers expected from this PR?  Yes.

### Issue(s) addressed

- fixes  [ufs-community/ccpp-physics/Issue #20](https://github.com/ufs-community/ccpp-physics/issues/20)


## Testing

How were these changes tested? What compilers / HPCs was it tested with? Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Have regression tests and unit tests (utests) been run? On which platforms and with which compilers? (Note that unit tests can only be run on tier-1 platforms)

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel
- [x] cheyenne.gnu
- [x] gaea.intel
- [x] jet.intel
- [x] wcoss2.intel
- [x] acorn.intel
- [ ] opnReqTest for newly added/changed feature
- [ ] CI

- Will the code updates change regression test baseline? Yes.  Changes to CCPP unified_ugwp drag suite. Please show the baseline directory below.
/scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/REGRESSION_TEST_INTEL.2022_11_08


## Dependencies

If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).

https://github.com/mdtoyNOAA/ufs-weather-model/tree/develop_ugwp_improvements
https://github.com/mdtoyNOAA/ccpp-physics/tree/ufs/dev_ugwp_improvements
https://github.com/mdtoyNOAA/fv3atm/tree/develop_ugwp_improvements

CCPP-physics pull request dependency: [ccpp-physics#20](https://github.com/ufs-community/ccpp-physics/pull/20)
FV3-ATM pull request dependency:  [fv3atm#600](https://github.com/NOAA-EMC/fv3atm/pull/600)

jkbk2004 commented:
@mdtoyNOAA Can you sync up branch to resolve the conflicts? We have open schedule today. If possible, we like to start working on this pr.

mdtoyNOAA commented:
I believe I am synced up.  I haven’t run any regression tests yet.  Will you take care of that?



> On Nov 29, 2022, at 12:37 PM, JONG KIM ***@***.***> wrote:
> 
> 
> @mdtoyNOAA <https://github.com/mdtoyNOAA> Can you sync up branch to resolve the conflicts? We have open schedule today. If possible, we like to start working on this pr.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-weather-model/pull/1488#issuecomment-1331194670>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARRVLIB7SOOGMXFUJGMYRTDWKZLPTANCNFSM6AAAAAAR44KAZE>.
> You are receiving this because you were mentioned.
>

jkbk2004 commented:
@BrianCurtis-NOAA @mdtoyNOAA I am starting to add the labels for hera.

BrianCurtis-NOAA commented:
***Automated RT Failure Notification***
Machine: hera
Compiler: gnu
Job: BL
[BL] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/1118181857/20221129210017/ufs-weather-model
Please make changes and add the following label back: hera-gnu-BL

BrianCurtis-NOAA commented:
***Automated RT Failure Notification***
Machine: hera
Compiler: intel
Job: BL
[BL] Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/1118181857/20221129210721/ufs-weather-model
Please make changes and add the following label back: hera-intel-BL

jkbk2004 commented:
hera/nems disk space issue. I cleaned up baselines and some people cleaned up their space as well. we are ok now 
/scratch1/NCEPDEV/nems DiskInUse=37402 GB, Quota=43000 GB

jkbk2004 commented:
@mdtoyNOAA do you have access to cheyenne? I see some failures on cheyenne gnu. It complains about FV3/ccpp/physics/physics/cires_ugwp.F90:327. Failure cases are control control_c48 control_stochy control_ras control_p8 rrfs_v1beta cpld_control_p8 cpld_control_nowave_noaero_p8

mdtoyNOAA commented:
No, I don't have access to cheyenne.  I don't know what would be causing
the failures.  The tests have in common that they use the "cires_ugwp" drag
scheme in their CCPP suites, except for test "control_p8", which doesn't
even call "cires_ugwp.F90".

On Tue, Nov 29, 2022 at 8:34 PM JONG KIM ***@***.***> wrote:

> @mdtoyNOAA <https://github.com/mdtoyNOAA> do you have access to cheyenne?
> I see some failures on cheyenne gnu. It complains about
> FV3/ccpp/physics/physics/cires_ugwp.F90:327. Failure cases are control
> control_c48 control_stochy control_ras control_p8 rrfs_v1beta
> cpld_control_p8 cpld_control_nowave_noaero_p8
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/pull/1488#issuecomment-1331601983>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARRVLIC7VHHYCHQJ23U4RGTWK3DLRANCNFSM6AAAAAAR44KAZE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
—————————
Michael Toy, Research Scientist
CIRES, University of Colorado Boulder
NOAA Global Systems Laboratory
325 Broadway, Boulder, Colorado
***@***.***
+1- <(303)%20497-4061>505-609-8782

jkbk2004 commented:
It looks like the code changes are leading to invalid memory reference issue along with that routine. @ChunxiZhang-NOAA @grantfirl @dustinswales Sounds like we need debugging on this. Hera/gnu runs ok but only on trouble on cheyenne/gnu. I double checked the develop branch is ok.

jkbk2004 commented:
@pjpegion FYI about the issue on this pr.

mdtoyNOAA commented:
@jkbk2004 It sounds like the gnu failures on cheyenne are occurring at run time (not during compilation).  Is that correct?   Also, it looks like the tests that fail happen to not be compiled in debug mode.  Their "debug" counterparts are passing?

jkbk2004 commented:
Yes, it's runtime crash. invalid memory reference... sounds like memory pointer issue or leaking

[err.txt](https://github.com/ufs-community/ufs-weather-model/files/10128055/err.txt)

dustinswales commented:
@mdtoyNOAA @jkbk2004 I'm looking into this on Cheyenne now.

dustinswales commented:
@jkbk2004 @mdtoyNOAA 
I checked out Mike's code and ran the offending RTs mentioned above and didn't run into any runtime errors?
[err.control_p8.OnCheyenneGNU.djs.txt](https://github.com/ufs-community/ufs-weather-model/files/10135336/err.control_p8.OnCheyenneGNU.djs.txt)

dustinswales commented:
@mdtoyNOAA @jkbk2004 
I was able to reproduce this issue on Cheyenne for the RTs listed above. Turns out a few [new input arguments to gwdps_run() needed to be declared as assumed shape](https://github.com/mdtoyNOAA/ccpp-physics/pull/1/files), since they are conditionally allocated in GFS_typedefs.
Mike, I opened a [PR with this change into your branch](https://github.com/mdtoyNOAA/ccpp-physics/pull/1)

jkbk2004 commented:
@dustinswales Thanks so much!

mdtoyNOAA commented:
@dustinswales  Great catch!  I accepted your PR and merged, pushed, etc.  Thank you very much.

jkbk2004 commented:
@mdtoyNOAA would it be ok to hold a bit more until all reviews finish? I think we are going to start to commit mixed-mode PR and fms update.

mdtoyNOAA commented:
> @mdtoyNOAA would it be ok to hold a bit more until all reviews finish? I think we are going to start to commit mixed-mode PR and fms update.

Sure, no problem.

mdtoyNOAA commented:
@jkbk2004  I've been periodically running RT's on the updated code -- first running "rt.sh -c" followed by "rt.sh -m".  The following "-m" tests are failing:
control_2dwrtdecomp
control_fhzero
control_2threads_debug

These are failing because of missing baseline directories.  These directories are not being created in REGRESSION_TEST by the "rt.sh -c" test:
REGRESSION_TEST/control
REGRESSION_TEST/control_debug

My testing log files are on Hera at:
/scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/tests

The REGRESSION_TEST directory is at:
/scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/REGRESSION_TEST

DeniseWorthen commented:
@mdtoyNOAA The latest PR to UWM (1525) fixed this issue. Please update and try again.

mdtoyNOAA commented:
> @mdtoyNOAA The latest PR to UWM (1525) fixed this issue. Please update and try again.

@DeniseWorthen   Oops, you're right.  I thought I had pulled the latest UWM version.  Thank you.

mdtoyNOAA commented:
I should note that the "control_p8" tests fails due to the updates to the calculation of "_u-gwd_ave_" and "_v-gwd_ave_" to include small-scale GWD and TOFD.  The "atmf" netCDF files are not effected by the change, so their comparisons to the baseline pass the test.  However, the "sfcf" netcdf files fail due to slight changes in these two diagnostic variables.  An "nccmp" comparison of these files gives:
> nccmp -dmfqS /scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/REGRESSION_TEST_INTEL.2022_11_08/control_p8/sfcf024.nc /scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/develop-20221205/INTEL/control_p8/sfcf024.nc
Variable  Group Count      Sum  AbsSum        Min       Max     Range         Mean     StdDev
u-gwd_ave /      8703 -1.28011 25.4602 -0.0135348 0.0162042  0.029739 -0.000147088 0.00418478
v-gwd_ave /      8359  6.60917 23.1362 -0.0103279 0.0141712 0.0244991  0.000790665 0.00394836

The important take-away is that this PR does not change the forecasts of the current control_p8 configuration.

ChunxiZhang-NOAA commented:
> I should note that the "control_p8" tests fails due to the updates to the calculation of "_u-gwd_ave_" and "_v-gwd_ave_" to include small-scale GWD and TOFD. The "atmf" netCDF files are not effected by the change, so their comparisons to the baseline pass the test. However, the "sfcf" netcdf files fail due to slight changes in these two diagnostic variables. An "nccmp" comparison of these files gives:
> 
> > nccmp -dmfqS /scratch1/BMC/wrfruc/mtoy/git_local/ufs-weather-model/REGRESSION_TEST_INTEL.2022_11_08/control_p8/sfcf024.nc /scratch1/NCEPDEV/nems/emc.nemspara/RT/NEMSfv3gfs/develop-20221205/INTEL/control_p8/sfcf024.nc
> > Variable  Group Count      Sum  AbsSum        Min       Max     Range         Mean     StdDev
> > u-gwd_ave /      8703 -1.28011 25.4602 -0.0135348 0.0162042  0.029739 -0.000147088 0.00418478
> > v-gwd_ave /      8359  6.60917 23.1362 -0.0103279 0.0141712 0.0244991  0.000790665 0.00394836
> 
> The important take-away is that this PR does not change the forecasts of the current control_p8 configuration.

@mdtoyNOAA Good to know. Thanks for the updates.

jkbk2004 commented:
@mdtoyNOAA we are working #1497. It's likely to merge in #1497 around noon eastern time today. Then, we can start working on this pr.

mdtoyNOAA commented:
I merged the latest FV3-atm changes and pushed to my fork.  There were no changes to the ufs-weather-model associated with #1497.  Will that be forthcoming?


> On Dec 20, 2022, at 7:13 AM, JONG KIM ***@***.***> wrote:
> 
> 
> @mdtoyNOAA <https://github.com/mdtoyNOAA> we are working #1497 <https://github.com/ufs-community/ufs-weather-model/pull/1497>. It's likely to merge in #1497 <https://github.com/ufs-community/ufs-weather-model/pull/1497> around noon eastern time today. Then, we can start working on this pr.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-weather-model/pull/1488#issuecomment-1359426319>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARRVLIGPFBXWSIJB6KLNH4LWOG5JBANCNFSM6AAAAAAR44KAZE>.
> You are receiving this because you were mentioned.
>

jkbk2004 commented:
@mdtoyNOAA we just merged in #1497.  One more time to sync up to resolve the conflicts. Then we will start testing.

mdtoyNOAA commented:
I just pulled the latest and pushed to my fork.  I think we’re synched up.



> On Dec 20, 2022, at 10:23 AM, JONG KIM ***@***.***> wrote:
> 
> 
> @mdtoyNOAA <https://github.com/mdtoyNOAA> we just merged in #1497 <https://github.com/ufs-community/ufs-weather-model/pull/1497>. One more time to sync up to resolve the conflicts. Then we will start testing.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-weather-model/pull/1488#issuecomment-1359850786>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARRVLIB7COWSD2LUUIDMYLLWOHTSTANCNFSM6AAAAAAR44KAZE>.
> You are receiving this because you were mentioned.
>

jkbk2004 commented:
@mdtoyNOAA can you check fv3atm hash? correct one is c6646f397b

mdtoyNOAA commented:
> @mdtoyNOAA can you check fv3atm hash? correct one is c6646f397b

My latest fv3atm hash is 482b9acae, which I merged from c6646f397b.  (See log below):

commit 482b9acae1603e4e5281fe5196a22db7a8a7488b (HEAD -> develop_ugwp_improvements, origin/develop_ugwp_improvements)
Merge: 6f2d334 c6646f3
Author: Michael Toy <michael.toy@noaa.gov>
Date:   Tue Dec 20 16:47:25 2022 +0000

    Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into develop_ugwp_improvements

commit c6646f397be5932fa18c2bd04eedc88ddee695ba (upstream/develop, upstream/HEAD)
Author: Samuel Trahan (NOAA contractor) <39415369+SamuelTrahanNOAA@users.noreply.github.com>
Date:   Tue Dec 20 08:31:58 2022 -0700

    Refactor surface restart logic in FV3GFS_io.F90 to not use hard-coded indices (#605)

jkbk2004 commented:
@mdtoyNOAA sounds like I was confused. I am adding new BL_DATE now.

jkbk2004 commented:
autoRT picked up PR ok and ran BL creation and RT. But logs were not pushed. It could be repository permission issue. @zach1221  can you check if log is pushed ok to branch in #1531 when you test autoRT thru Jenkins ?

ChunxiZhang-NOAA commented:
Many files are changed in the tests directory, such as the regression tests for rap.  Did this PR change these files?

zach1221 commented:
@jkbk2004 testing now on #1531

BrianCurtis-NOAA commented:
> autoRT picked up PR ok and ran BL creation and RT. But logs were not pushed. It could be repository permission issue. @zach1221 can you check if log is pushed ok to branch in #1531 when you test autoRT thru Jenkins ?

This is something that you should address with something along the lines of a "bot" account. This way keys can be created from that account to use in AutoRT. Then the bots access can be set such that it will be allowed to push to PR branches. Of course with such heightened permissions, you'd want to limit who has credentials to the account.

BrianCurtis-NOAA commented:
@mdtoyNOAA Is there a reason there are a bunch of changes to regression tests? I didn't see it mentioned in the PR.

jkbk2004 commented:
@mdtoyNOAA Even though we see the change inrap.nml.IN, all the baselines get affected by ugwp update/change, right?

jkbk2004 commented:
@mdtoyNOAA This PR contains extensive changes on tests files to switch from NEW_DIAGTABLE to DIAG_TABLE. I tried to do some sanity check on these changes. I think we need to confirm  about those changes as well. I think we need more time to work on this pr. @BrianCurtis-NOAA  I will move on to #1523.

mdtoyNOAA commented:
@BrianCurtis-NOAA @jkbk2004 The changes to the regression tests are mentioned in the description (quoted below).  These involve changing "NEW_DIAGTABLE" (obsolete) to "DIAG_TABLE".

_"Also, fixed bugs in the rap, hrrr and rrfs test descriptions (changed instances of NEW_DIAGTABLE to DIAG_TABLE) and deleted extraneous line in tests/parm/rap.nml.IN (i.e., ldiag_ugwp=.false.)."_

mdtoyNOAA commented:
> @mdtoyNOAA Even though we see the change inrap.nml.IN, all the baselines get affected by ugwp update/change, right?

@jkbk2004 Yes, that is correct.

mdtoyNOAA commented:
@BrianCurtis-NOAA @jkbk2004 The switch from NEW_DIAGTABLE to DIAG_TABLE in the "rap" test files was a trivial bug fix, which was brought to my attention about a month ago.  Somewhere along the way, "NEW_DIAGTABLE" had become obsolete and replaced by "DIAG_TABLE".  For some reason, when this change was made, the "rap" test files were overlooked, and were not updated.  This PR updates these tests, and I had added the note in the description describing this.

jkbk2004 commented:
I agree. I was confused because I see no other environment variables depend on NEW_DIAGTABLE in develop branch. @SamuelTrahanNOAA @RatkoVasic-NOAA can we clarify about the issue?

jkbk2004 commented:
As @mdtoyNOAA points out. I see NEW_DIAGTABLE is a terminal variable. No other variables depend on NEW_DIAGTABLE. Then, question moves to accept baselines changes. I remember Mike attached a few figures to validate the result. @mdtoyNOAA can we re-confirm?

SamuelTrahanNOAA commented:
The NEW_DIAGTABLE was an error, and I think the result is that it was using the DIAG_TABLE specified in default_vars.sh.

jkbk2004 commented:
@SamuelTrahanNOAA thanks for the comment! @mdtoyNOAA once you are happy with change result, we are good to go for final approvals.

mdtoyNOAA commented:
> @SamuelTrahanNOAA thanks for the comment! @mdtoyNOAA once you are happy with change result, we are good to go for final approvals.

@jkbk2004  I know that the NEW_DIAGTABLE to DIAG_TABLE fix did not change the results.  So I agree on getting final approvals.

jkbk2004 commented:
@BrianCurtis-NOAA I think new_diagtable issue is resolved. Would it be ok to skip the baseline creation and test on wcoss2?

BrianCurtis-NOAA commented:
> @BrianCurtis-NOAA I think new_diagtable issue is resolved. Would it be ok to skip the baseline creation and test on wcoss2?

No. Let me try finishing things by tomorrow afternoon.

jkbk2004 commented:
> > @BrianCurtis-NOAA I think new_diagtable issue is resolved. Would it be ok to skip the baseline creation and test on wcoss2?
> 
> No. Let me try finishing things by tomorrow afternoon.

Sure!

BrianCurtis-NOAA commented:
The Acorn.Intel log is on Hera at /scratch1/NCEPDEV/stmp2/Brian.Curtis/ @mdtoyNOAA please add to your PR branch.

BrianCurtis-NOAA commented:
> The Acorn.Intel log is on Hera at /scratch1/NCEPDEV/stmp2/Brian.Curtis/ @mdtoyNOAA please add to your PR branch.

@mdtoyNOAA The WCOSS2.Intel log is also on hera at the same location. Please add both to your PR branch.

jkbk2004 commented:
@BrianCurtis-NOAA @mdtoyNOAA I was able to push acorn and wcoss2 logs from url. Sounds like all sets.

jkbk2004 commented:
@ChunxiZhang-NOAA we can start merging in https://github.com/ufs-community/ccpp-physics/issues/20

ChunxiZhang-NOAA commented:
@jkbk2004 Done.

jkbk2004 commented:
@mdtoyNOAA fv3atm pr was merged. can you update submodule to that and revert branch in gitmodules?

mdtoyNOAA commented:
Done.


> On Dec 22, 2022, at 10:05 AM, JONG KIM ***@***.***> wrote:
> 
> 
> @mdtoyNOAA <https://github.com/mdtoyNOAA> fv3atm pr was merged. can you update submodule to that and revert branch in gitmodules?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-weather-model/pull/1488#issuecomment-1363107625>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARRVLIE73KVH72KYY3IA4NDWOSC4VANCNFSM6AAAAAAR44KAZE>.
> You are receiving this because you were mentioned.
>

jkbk2004 commented:
looks good!
